app/models/events/tags.rb

```diff
@@ -0,0 +1,39 @@
+module Events::Tags
+  extend ActiveSupport::Concern
+
+  included do
+    validates :tag, presence: true, inclusion: { in: ->(_) { tags.keys } }
+    attribute :tag, Events::Tag::Type.new
+    validates :additional, nested: true, allow_nil: true
+    belongs_to :ip_address, optional: true
+    belongs_to :geoip_info, optional: true
+
+    cattr_reader(:tags) { {} }
+
+    def additional_type
+      tags.fetch(tag, nil)
+    end
+
+    def additional
+      additional_type&.new(super) || super
+    end
+
+    def additional=(value)
+      super(value&.to_h)
+    end
+  end
+
+  class_methods do
+    def define_event(string, &blk)
```

Comment:
Maybe some of this method would be more local in Event::Tag, since it's almost all about that class?

Reply:
I'm not really seeing what else can be pushed there, since we're dealing with the `tags` method and setting consts on `self`